api/agent/func_logger.go

```diff
@@ -148,8 +120,9 @@ func (li *lineWriter) Write(ogb []byte) (int, error) {
 		// write in this line and advance buffer past it
 		l := b[:i+1]
 		ns, err := li.w.Write(l)
+		n += ns
```

Comment:
Write() should return "how much of the original buffer was consumed". Looks like existing bug... In other words, in this state, Write() may return n, which can be larger than len(ogb) which is not really what we want to return.. I think we might want to change a few things such as:

- li.b buffer should be limited. Currently, series of Write() calls without a newline can grow li.b.buffer without limits with each Write() having to parse entire buf for newlines.
- if buffer is at limit (or going to grow over limit with current call to Write()), then we should flush li.b with appended newline ourselves before processing ogb. If that flush fails, then we can return error.

Reply:
i guess, in practice, the reality of this is that the inner writer is never going to error (it's a logWriter, which never errors) and writing to/reading from the buffer is never going to error either (we're not setting a capacity on buffers, and many methods of the buffer contract will never error - the runtime will oom before an error gets returned, but this is a panic). it's also the case that this is not to be used for production applications and we can avoid thinking about some of the more fun attack vectors here, but would be happy to note them - I just don't think we need to add any more arbitrary configuration than we already have for something that has a configuration to just turn it off for prod apps (glad to fix correctness issues, of course).

that being said, appreciate the feedback can spend an hour here tidying this up so that it's more obviously correct.